### DISCUSSION TITLE
[Q] How to use the capture from a first pattern as a predicate for a second one?

### DISCUSSION THREAD

**Discussion by gagbo** [tree-sitter/tree-sitter#2318] [Q&A]
It is somewhat linked to #981 

For context, I put everything in that Gist with a screenshot, a repro, and everything: https://gist.github.com/gagbo/f9e9ff04633391408064068356bca4c6

Since I can't match elements at any nested level, I'm trying to make a query with 2 patterns, where the first pattern detects an import name ("autometrics"), and the second pattern only matches nodes based on that first pattern detection.

Currently when I'm trying the pattern, I get a false positive (on line 7 of the sample code), where the name of the function in the `call_expression` doesn't match the name that got saved in the earlier pattern.

Is there something I'm missing to make this work, or am I obligated to run this as 2 separate query steps ? (with templating to fill the second query with the result of the first)

**Comment by sogaiu** [comment]
I don't have a whole lot of helpful bits to contribute, but FWIW:

* Tried things out and noticed that the version of tree-sitter-javascript at the "official" playground doesn't appear to understand `string_fragment`
* I ran a local version of the playground with the latest for tree-sitter-javascript and replicated the reported results
* IIUC, `#eq?` and friend predicates are all "re"-implemented per language binding so the behavior that's being described here may be specific to the web / wasm bindings (`lib/binding_web`) -- at least when observing via the playground.  May be that makes [this code](https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/binding.js#L772-L803) relevant?

The [docs here](https://tree-sitter.github.io/tree-sitter/using-parsers#predicates) suggest that using more than one capture name with a predicate is supposed to work, but it's not clear to me whether a capture name from one pattern is supposed to work inside another pattern.  (But note the point above about the predicates being "per-binding".)

**Reply by gagbo** [reply]
I think you need to use the tree-sitter-_typescript_ bindings. Even with a somewhat old version (0.12.something) on my local setup I had the string_fragment in the nodes, and I also had it in my 0.20.something version on my other local setup.

But all the other points are still relevant, thanks :)

I know the predicates are implemented per language binding, and I thought that all bindings at least implemented `#eq?`; but maybe they have different implementations for the case where we want to compare 2 captures

**Reply by sogaiu** [reply]
> I think you need to use the tree-sitter-typescript bindings.

Oh, oops :)

Thanks for pointing that out.

**Reply by sogaiu** [reply]
FWIW, when I tried with a local version of the typescript grammar, and `(#eq? @wrapperAtCall "autometrics")` substituted in for `(#eq? @wrapperAtCall @wrapperAtName)`, `req` and `res` are not captured, while the other desired captures are.

When I leave out `(#eq? @wrapperAtCall @wrapperAtName)` completely, I get the results described in the gist (so `req` and `res` are captured as well as the 3 other things on other lines).

Perhaps `(#eq? @wrapperAtCall @wrapperAtName)` is just being ignored?

**Reply by gagbo** [reply]
Maybe the whole predicate is being ignored, but then I suppose it would be a bug in the TS bindings, since it seems to have special code to deal with `#eq?` when the [2nd operand is a capture](https://github.com/tree-sitter/tree-sitter/blob/49226023693107fba9a1191136a4f47f38cdca73/lib/binding_web/binding.js#L779-L790).

If the reason my predicate is not effective is that `@wrapperAtName` is not "defined" in the 2nd pattern, then I'd like to know what the workaround @ahlinc [talked about in 981](https://github.com/tree-sitter/tree-sitter/discussions/981#discussioncomment-561985) is, because I can't figure it out

**Reply by sogaiu** [reply]
> it seems to have special code to deal with #eq? when the [2nd operand is a capture](https://github.com/tree-sitter/tree-sitter/blob/49226023693107fba9a1191136a4f47f38cdca73/lib/binding_web/binding.js#L779-L790).

Yes, it does look like there is some code to deal with the 2nd operand being a capture -- but may be that only applies if both captures are part of the same pattern?

To quote ahlinc:

> you can workaround this by writing a query to find all class_declaration nodes and after that on every class declaration node run a query that would match all identifiers in invocation expressions.

That seems like two queries to me.  Or am I misunderstanding?
